File: DSAlgorithms/Leetcode/c++/409-longestPalindrome.cpp
#include <iostream>
#include <unordered_map>


int longestPalindrome(std::string );
int main()
{
    std::string str = "abccccdd";
    std::cout<<longestPalindrome(str);
    return 0;
}
int longestPalindrome(std::string s) {
        int sum = 0;    
        if(s.size() == 0)
            return sum;
        std::unordered_map<char, int> hash;
        int string_size = s.size();
        for(int i =0; i < string_size; i++)
        {
            hash[s[i]]++;
        }
        for(std::unordered_map<char,int>::iterator it = hash.begin(); it != hash.end(); ++it)
        {
            sum+= (it->second/2) * 2;  //int division will throw out decimal 1/2 -> 0 4/5-> 2
        }
        if(string_size > sum)
            return ++sum;
        return sum;
    }

File: DSAlgorithms/Leetcode/ts/src/409-longestPalindrome.ts
const longestPalindrome = (s: string): number => {
    let sum = 0;
    const string_size = s.length;
    if(string_size === 0)
        return sum;
    const hash = new Map<string, number>();
    for(let  i = 0; i < string_size; i++)
        {
            if(hash.has(s[i]))
                {
                    let count = hash.get(s[i]);
                    if(count)
                        hash.set(s[i], ++count)
                }
            else
                hash.set(s[i], 1);
        }
    hash.forEach((value:number, key: string) => {
        sum+= Math.floor(value/2) * 2;
    })
    return sum < string_size ? ++sum : sum;
};

console.log(longestPalindrome("abcccdd"));